libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -0,0 +1,801 @@
+/*
+ * AP_UAVCAN.cpp
+ *
+ *      Author: Eugene Shamaev
+ */
+
+#include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_WITH_UAVCAN
+
+#include "AP_UAVCAN.h"
+#include <GCS_MAVLink/GCS.h>
+
+// Zubax GPS and other GPS, baro, magnetic sensors
+#include <uavcan/equipment/gnss/Fix.hpp>
+#include <uavcan/equipment/gnss/Auxiliary.hpp>
+#include <uavcan/equipment/ahrs/MagneticFieldStrength.hpp>
+#include <uavcan/equipment/air_data/StaticPressure.hpp>
+#include <uavcan/equipment/air_data/StaticTemperature.hpp>
+#include <uavcan/equipment/actuator/ArrayCommand.hpp>
+#include <uavcan/equipment/actuator/Command.hpp>
+#include <uavcan/equipment/actuator/Status.hpp>
+#include <uavcan/equipment/esc/RawCommand.hpp>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define debug_uavcan(level, fmt, args...) do { if ((level) <= AP_BoardConfig::get_can_debug()) { hal.console->printf(fmt, ##args); }} while (0)
+
+// Translation of all messages from UAVCAN structures into AP structures is done
+// in AP_UAVCAN and not in corresponding drivers.
+// The overhead of including definitions of DSDL is very high and it is best to
+// concentrate in one place.
```

Comment:
The reason for that is that libuavcan instantiates a helper object per every inclusion of a generated data type header that registers certain information about the data type in the library at runtime. The helper object is defined here: https://github.com/UAVCAN/libuavcan/blob/master/libuavcan/dsdl_compiler/libuavcan_dsdl_compiler/data_type_template.tmpl#L453. Its data type is defined here: https://github.com/UAVCAN/libuavcan/blob/master/libuavcan/include/uavcan/node/global_data_type_registry.hpp#L177-L237. There is a library configuration option, `UAVCAN_NO_GLOBAL_DATA_TYPE_REGISTRY`, that disables this feature, but it makes the developer's life a lot harder.